HIP/hip-729.md

```diff
@@ -0,0 +1,117 @@
+---
+hip: 729
+title: Contract Accounts Nonces Externalization
+author: Miroslav Gatsanoga <miroslav.gatsanoga@limechain.tech>
+working-group: Stoyan Panayotov <stoyan.panayotov@limechain.tech>, Ivan Kavaldzhiev <ivan.kavaldzhiev@limechain.tech>, Richard Bair <@rbair23>, Nana Essilfie-Conduah <@Nana-EC>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+created: 2023-05-05
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/730
+updated: 2023-05-09
+---
+
+## Abstract
+
+The Hedera Smart Contract Service (HSCS) supports the deployment of contracts. A deployed contract should have a nonce value that depends on the number of other contracts that were created during its own creation.
+Currently this is not the observed behavior in Hedera and this HIP aims to fix that.
+
+## Motivation
+
+In Ethereum, contract accounts have nonces, as specified by the [yellow paper](http://gavwood.com/paper.pdf) and [EIP-161](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-161.md).
+
+The EVM uses the nonce value internally, for several different operations, for example:
+    1. to check whether an account exists at an address, it checks whether the nonce property stored for this address is different than 0 (it also checks the balance and code properties for the address as well, but this is out of scope for this discussion).
+    2. when deploying new contracts via a `CREATE` operation, the address for an Ethereum contract is **deterministically** computed from the address of its creator (`sender`) and how many transactions the creator has sent (`nonce`). The `sender` and `nonce` are RLP encoded and then hashed with Keccak-256.
+
+Ethereum nodes expose a JSON-RPC endpoint that can be queried for each account’s nonce value: [eth_getTransactionCount](https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_gettransactioncount)
+
+In Hedera there is `JSON-RPC Relay` (as described in [HIP-482](https://hips.hedera.com/hip/hip-482)) that provides the same `eth_getTransactionCount` JSON-RPC endpoint, but currently it always returns a value of 1 for contract accounts. The desired behavior would be to have a nonce value of `1 + n` where `n` is the number of times the contract created other contracts through `CREATE` or `CREATE2` operations.
```

Comment:
Updated to highlight where the data is wrong and the chain effect.
```suggestion
In Hedera the Mirror Node which stores historical entity and transaction information has no knowledge of contract nonces and always returns a 0 value. As a result, the `JSON-RPC Relay` (as described in [HIP-482](https://hips.hedera.com/hip/hip-482)) which exposes the `eth_getTransactionCount` JSON-RPC endpoint has to query the Consensus node and not the Mirror Node to return nonce values for contract accounts. The desired behavior would be to have a nonce value of `1 + n` where `n` is the number of times the contract created other contracts through `CREATE` or `CREATE2` operations exposed to the Mirror Node.
```

Reply:
Updated.